File: src/main/java/br/com/augustoccesar/querybuilder/query/Column.java
package br.com.augustoccesar.querybuilder.query;

import br.com.augustoccesar.querybuilder.builders.Buildable;
import br.com.augustoccesar.querybuilder.constants.CommonStrings;
import br.com.augustoccesar.querybuilder.exceptions.InvalidPattern;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by augustoccesar on 8/9/16.
 */
public class Column implements Buildable{
    private String prefix;
    private String name;
    private String alias;
    private boolean distinct = false;

    // Constructors

    public Column(String name, boolean distinct) {
        this.name = name;
        this.distinct = distinct;
    }

    public Column(String prefix, String name, boolean distinct) {
        this.prefix = prefix;
        this.name = name;
        this.alias = prefix + "_" + name;
        this.distinct = distinct;
    }

    public Column(String prefix, String name, String customAlias, boolean distinct) {
        this.name = name;
        this.prefix = prefix;
        this.alias = customAlias;
        this.distinct = distinct;
    }

    // Builder

    public static Column fromMarkdown(String columnMarkdown) {
        Pattern markdownWithTableAlias = Pattern.compile("\\{(\\w+)\\}(\\**\\w+)");
        Pattern fullMarkdown = Pattern.compile("\\{(\\w+)\\}(\\**\\w+)\\{(\\w+)\\}");

        Matcher markdownWithTableAliasMatcher = markdownWithTableAlias.matcher(columnMarkdown);
        Matcher fullMarkdownMatcher = fullMarkdown.matcher(columnMarkdown);

        if(markdownWithTableAliasMatcher.matches()){
            String prefix = markdownWithTableAliasMatcher.group(1);
            String name = markdownWithTableAliasMatcher.group(2);
            boolean distinct = false;

            if (name.contains("*")){
                name = name.replace("*", "");
                distinct = true;
            }

            if("_".equals(prefix)){
                return new Column(name, distinct);
            }else {
                return new Column(prefix, name, distinct);
            }
        }else if(fullMarkdownMatcher.matches()){
            String prefix = fullMarkdownMatcher.group(1);
            String name = fullMarkdownMatcher.group(2);
            String customAlias = fullMarkdownMatcher.group(3);
            boolean distinct = false;

            prefix = "_".equals(prefix) ? null : prefix;
            customAlias = "_".equals(customAlias) ? null : customAlias;

            if (name.contains("*")){
                name = name.replace("*", "");
                distinct = true;
            }

            return new Column(prefix, name, customAlias, distinct);
        }else{
            throw new InvalidPattern("Column");
        }
    }

    // Getters and Setters

    public String getName() {
        return name;
    }

    public String getPrefix() {
        return prefix;
    }

    public String getAlias() {
        return alias;
    }

    public boolean isDistinct() {
        return distinct;
    }

    @Override
    public String build() {
        StringBuilder stringBuilder = new StringBuilder();

        if(this.distinct) {
            stringBuilder.append(CommonStrings.DISTINCT);
        }

        if(this.prefix != null){
            stringBuilder.append(this.prefix).append(".");
        }

        stringBuilder.append(this.name);

        if(this.alias != null){
            stringBuilder.append(CommonStrings.AS).append(this.alias);
        }

        return stringBuilder.toString();
    }

    @Override
    public String build(boolean withDistinct, boolean withAlias) {
        StringBuilder stringBuilder = new StringBuilder();

        if(withDistinct) {
            stringBuilder.append(CommonStrings.DISTINCT);
        }

        if(this.prefix != null){
            stringBuilder.append(this.prefix).append(".");
        }

        stringBuilder.append(this.name);

        if(withAlias){
            stringBuilder.append(CommonStrings.AS).append(this.alias);
        }

        return stringBuilder.toString();
    }
}

File: src/main/java/br/com/augustoccesar/querybuilder/query/Column.kt
package br.com.augustoccesar.querybuilder.query

import br.com.augustoccesar.querybuilder.builders.Buildable
import br.com.augustoccesar.querybuilder.constants.CommonStrings
import br.com.augustoccesar.querybuilder.exceptions.InvalidPattern
import java.util.regex.Pattern

/**
 * Created by augustoccesar on 8/9/16.
 */
class Column(var name: String) : Buildable {
    var prefix: String? = null
    var alias: String? = null
    var distinct: Boolean = false

    constructor(name: String, distinct: Boolean) : this(name) {
        this.distinct = distinct
    }

    constructor(prefix: String, name: String, distinct: Boolean) : this(name) {
        this.prefix = prefix
        this.alias = prefix + "_" + name
        this.distinct = distinct
    }

    constructor(prefix: String?, name: String, alias: String?, distinct: Boolean) : this(name) {
        this.prefix = prefix
        this.alias = alias
        this.distinct = distinct
    }

    fun sqlColumnRepresentation(): String {
        val stringBuilder = StringBuilder()
        if (prefix != null)
            stringBuilder.append(prefix).append(".")
        stringBuilder.append(name)
        return stringBuilder.toString()
    }

    fun isDistinct(): Boolean {
        return distinct
    }

    override fun build(): String {
        val stringBuilder = StringBuilder()

        if (this.distinct) {
            stringBuilder.append(CommonStrings.DISTINCT)
        }

        if (this.prefix != null) {
            stringBuilder.append(this.prefix).append(".")
        }

        stringBuilder.append(this.name)

        if (this.alias != null) {
            stringBuilder.append(CommonStrings.AS).append(this.alias)
        }

        return stringBuilder.toString().replace("\\s+".toRegex(), " ")
    }

    companion object {
        @JvmStatic
        fun multiColumns(tablePrefix: String, vararg columnMarkdown: String): MutableList<String> {
            val response: MutableList<String> = mutableListOf()
            val regex = "(\\**\\w+)(\\{(\\w+)\\})?".toRegex()

            for(i in 0..columnMarkdown.size - 1) {
                val groups = regex.matchEntire(columnMarkdown[i])?.groups ?: throw InvalidPattern("Column")

                val name = groups[1]?.value
                val alias = groups[3]?.value

                var fullMarked: String = "{$tablePrefix}$name"
                if(alias != null) fullMarked = "$fullMarked{$alias}"

                response.add(fullMarked)
            }

            return response
        }

        @JvmStatic
        fun fromMarkdown(columnMarkdown: String): Column {
            val markdownWithTableAlias = Pattern.compile("\\{(\\w+)\\}(\\**\\w+)")
            val fullMarkdown = Pattern.compile("\\{(\\w+)\\}(\\**\\w+)\\{(\\w+)\\}")

            val markdownWithTableAliasMatcher = markdownWithTableAlias.matcher(columnMarkdown)
            val fullMarkdownMatcher = fullMarkdown.matcher(columnMarkdown)

            if (markdownWithTableAliasMatcher.matches()) {
                val prefix = markdownWithTableAliasMatcher.group(1)
                var name = markdownWithTableAliasMatcher.group(2)
                var distinct = false

                if (name.contains("*")) {
                    name = name.replace("*", "")
                    distinct = true
                }

                if ("_" == prefix) {
                    return Column(name = name, distinct = distinct)
                } else {
                    return Column(prefix = prefix, name = name, distinct = distinct)
                }
            } else if (fullMarkdownMatcher.matches()) {
                var prefix: String? = fullMarkdownMatcher.group(1)
                var name = fullMarkdownMatcher.group(2)
                var customAlias: String? = fullMarkdownMatcher.group(3)
                var distinct = false

                prefix = if ("_" == prefix) null else prefix
                customAlias = if ("_" == customAlias) null else customAlias

                if (name.contains("*")) {
                    name = name.replace("*", "")
                    distinct = true
                }

                return Column(prefix, name, customAlias, distinct)
            } else {
                throw InvalidPattern("Column")
            }
        }
    }
}
